crates/bevy_render/src/render_resource/bindless.rs

```diff
@@ -0,0 +1,313 @@
+
+use super::binding_types::{
+    sampler, texture_1d, texture_2d, texture_2d_array, texture_3d, texture_cube, texture_cube_array,
+};
+
+/// The default value for the number of resources that can be stored in a slab
+/// on this platform.
+///
+/// See the documentation for [`BindlessSlabResourceLimit`] for more
+/// information.
+#[cfg(any(target_os = "macos", target_os = "ios"))]
+pub const AUTO_BINDLESS_SLAB_RESOURCE_LIMIT: u32 = 64;
+/// The default value for the number of resources that can be stored in a slab
+/// on this platform.
+///
+/// See the documentation for [`BindlessSlabResourceLimit`] for more
+/// information.
+#[cfg(not(any(target_os = "macos", target_os = "ios")))]
+pub const AUTO_BINDLESS_SLAB_RESOURCE_LIMIT: u32 = 2048;
+
+/// The binding numbers for the built-in binding arrays of each bindless
+/// resource type.
+///
+/// In the case of materials, the material allocator manages these binding
+/// arrays.
+///
+/// `bindless.wgsl` contains declarations of these arrays for use in your
+/// shaders. If you change these, make sure to update that file as well.
+pub static BINDING_NUMBERS: [(BindlessResourceType, BindingNumber); 9] = [
+    (BindlessResourceType::SamplerFiltering, BindingNumber(1)),
+    (BindlessResourceType::SamplerNonFiltering, BindingNumber(2)),
+    (BindlessResourceType::SamplerComparison, BindingNumber(3)),
+    (BindlessResourceType::Texture1d, BindingNumber(4)),
+    (BindlessResourceType::Texture2d, BindingNumber(5)),
+    (BindlessResourceType::Texture2dArray, BindingNumber(6)),
+    (BindlessResourceType::Texture3d, BindingNumber(7)),
+    (BindlessResourceType::TextureCube, BindingNumber(8)),
+    (BindlessResourceType::TextureCubeArray, BindingNumber(9)),
+];
+
+/// The maximum number of resources that can be stored in a slab.
+///
+/// This limit primarily exists in order to work around `wgpu` performance
+/// problems involving large numbers of bindless resources. Also, some
+/// platforms, such as Metal, currently enforce limits on the number of
+/// resources in use.
+///
+/// This corresponds to `LIMIT` in the `#[bindless(LIMIT)]` attribute when
+/// deriving [`crate::render_resource::AsBindGroup`].
+#[derive(Clone, Copy, Default, PartialEq, Debug)]
+pub enum BindlessSlabResourceLimit {
+    /// Allows the renderer to choose a reasonable value for the resource limit
+    /// based on the platform.
+    ///
+    /// This value has been tuned, so you should default to this value unless
+    /// you have special platform-specific considerations that prevent you from
+    /// using it.
+    #[default]
+    Auto,
+
+    /// A custom value for the resource limit.
+    ///
+    /// Bevy will allocate no more than this number of resources in a slab,
+    /// unless exceeding this value is necessary in order to allocate at all
+    /// (i.e. unless the number of bindless resources in your bind group exceeds
+    /// this value), in which case Bevy can exceed it.
```

Comment:
Does this exception just mean that in the cases where more bindings are needed than in this limit, we allow it and accept whatever the probably minimal performance hit is for this one material binding?

Reply:
More or less, yes. I don't really consider it a performance hit per se because the alternative is to not be able to render at all.